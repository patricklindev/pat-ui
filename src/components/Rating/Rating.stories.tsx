import React from 'react';
import Rating from './Rating';
import { action } from '@storybook/addon-actions';

export default {
    title: 'Rating',
    component: Rating,
  };

  export const DefaultRating = () => (
    <div>
      <div>Users  should be able to preview the rating by hovering on the ‘stars’(proper animation should be applied)</div>
      <Rating />
    </div>
  );
  
  export const DisabledRating = () => (
    <div>
      <div>Developers should be able to disable the component from props</div>
      <Rating edit={false}/>
    </div>
  );

  const labels = ['Very bad','Bad','Okay','Good','Great'];
  export const ProvideTheLabelForRating = () => (
    <div>
      <div>Developers should be able to provide the label of the component from props</div>
      <Rating labels = {labels}/>
    </div>
  );

  export const SetThePrecisionOfTheRating = () => (
    <div>
      <div>Developers should be able to set the precision of the rating component(Users can give a fraction of stars)</div>
      <Rating half={true}/>
    </div>
  );

  export const ChooseSizesOfTheRating = () => (
    <div>
      <div>Developers can choose sizes of the rating component among various predefined options from props</div>
      <br />
      <div>30px</div>
      <Rating size={30}/>
      <div>35px</div>
      <Rating size={35}/>
      <div>40px</div>
      <Rating size={40}/>
    </div>
  );

  export const DecideTheNumberOfStars = () => (
    <div>
      <div>Developers can decide the number of stars in total from props</div>
      <Rating count={3}/>
      <Rating count={6}/>
      <Rating count={9}/>
    </div>
  );

  export const ControlTheValueOfTheRating = () => (
    <div>
      <div>Developers should be able to control the value of the rating from outside of the component by providing a prop.</div>
      <Rating value={1}/>
      <Rating half={true} value={2.5}/>
      <Rating value={5}/>
    </div>
  );

  export const ListenToTheChangeOfTheRalue = () => (
    <div>
      <div>Developers should be able to listen to the change of the value of the component from outside of the component by providing the onChange callback function as a prop.</div>
      <Rating half onChange={action('New Rating is')}/>
    </div>
  );
  