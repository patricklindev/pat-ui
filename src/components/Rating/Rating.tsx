import React from 'react';

const Rating = () => {
  return <div>Rating</div>;
};

export default Rating;
