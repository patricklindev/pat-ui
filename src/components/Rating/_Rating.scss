.rating {
  display: inline-block;
  cursor: pointer;
  font-size: 0.9rem;

  svg {
    fill: $yellow;
    width: 23px;
  }

  &__stars {
    display: inline-block;
    position: relative;
    height: 100%;
  }
  &__label {
    margin-left: 25px;
  }

  &__stars-cover {
    position: absolute;
    height: 100%;
    background-color: white;
    right: 0;
    top: 0;
    mix-blend-mode: color; //blend element's background with its parent's background color
  }
}

.rating.disabled {
  cursor: not-allowed;
  svg {
    fill: yellow;
  }
}

.rating.rating-sm {
  svg {
    width: 18px;
  }
  font-size: 0.8rem;
}

.rating.rating-lg {
  svg {
    width: 30px;
  }
  font-size: 1rem;
}
