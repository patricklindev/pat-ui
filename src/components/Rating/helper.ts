import { IconColor, IconSize } from '../Icon/Icon';

export const getIconName = (starValue: number): string => {
  switch (starValue) {
    case 0:
      return 'star regular';
    case 0.5:
      return 'star half';
    default:
      return 'star';
  }
};

export const getIconColor = (starValue: number): IconColor => {
  if (starValue !== 0) {
    return 'orange';
  }
  return 'grey';
};

export const getSizeName = (size: IconSize, isHovering: boolean): IconSize => {
  switch (size) {
    case 'small':
      if (isHovering) {
        return 'medium';
      }
      return size;

    case 'medium':
      if (isHovering) {
        return 'large';
      }
      return size;

    case 'large':
      if (isHovering) {
        return 'big';
      }
      return size;

    default:
      return size;
  }
};

export const getSizeNumberByName = (size: IconSize): number => {
  switch (size) {
    case 'small':
      return 20;
    case 'medium':
      return 28;
    case 'large':
      return 42;
    default:
      return 28;
  }
};
