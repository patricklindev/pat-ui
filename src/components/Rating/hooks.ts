import { useEffect, useState } from 'react';

// rating label will change according to currentRating
export const useCurrentRatingLabel = (
  currentRating: number,
  ratingCount: number
) => {
  const [labelName, setLabelName] = useState('');

  const lowTile = Math.round((ratingCount as number) / 4);
  const midTile = Math.round((ratingCount as number) / 2);

  useEffect(() => {
    if (currentRating <= lowTile) {
      setLabelName('Poor');
    } else if (currentRating <= midTile) {
      setLabelName('Good');
    } else {
      setLabelName('Excellent!');
    }
  }, [currentRating, ratingCount]);

  return {
    labelName,
  };
};
