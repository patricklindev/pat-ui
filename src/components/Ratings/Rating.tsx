import React, { useState, useEffect, FC } from 'react';
import StarIcon from './StarIcon';

export interface IRatingProps {
    /**set true or false*/
    disableHandler?: boolean;
    /**set "sm" | "md" | "lg" */
    starSize?: string;
    /**set custom classname by sending in a prop*/
    classNames?: string;
    /**set star count by using array [1,2,3,4,5]*/
    count?: number;
    /**set customized default value of star. Example: 3*/
    customRatingValue?: number;
    /**set true or false to activate label on hover*/
    hoverLabel?: boolean;
    /**set custom icon size by sending prop in. Example: "75px" */
    customSize?: string;
    /**set true or false to activate custom label on hover*/
    customLabel?: boolean;
    /**set your own custom hover label text*/
    customHoverLabel?: any;
    /**send addon-action from storybook as a prop */
    onChange?: () => any;
}

export type patRatingProps = IRatingProps;

export const Rating: FC<patRatingProps> = (props: any) => {
    const {
        disableHandler,
        starSize,
        classNames,
        count,
        customRatingValue,
        hoverLabel,
        customSize,
        onChange,
        customHoverLabel,
        customLabel,
    } = props;

    const [rating, setRating] = useState<number>(customRatingValue);
    const [hoverRating, setHoverRating] = useState<number>(0);
    const [disableCss, setDisableCss] = useState<string>('');
    const [ratingLabel, setRatingLabel] = useState<string>('');
    const [customRatingLabel, setCustomRatingLabel] = useState('');

    //default hover label
    const starLabel: any = {
        1: 'Useless',
        2: 'Poor',
        3: 'OK',
        4: 'Good',
        5: 'Excellent',
    };

    //custom hover labels
    const customStarLabel: any = customHoverLabel;

    //add color on mouse enter
    const highlite = (index: number) => {
        setHoverRating(index);
    };

    //remove color on mouse leave
    const removeHighlite = () => {
        setHoverRating(0);
        removeHoverLabel();
    };

    //save color/rating on click
    const saveRating = (index: number) => {
        setRating(index);
    };

    //if disabled add disabled css
    useEffect(() => {
        if (disableHandler) {
            setDisableCss('disabled');
        }
    });

    //loop through starlabel or customStarLabel object to grab label value
    useEffect(() => {
        if (hoverLabel) {
            for (let i = 0; i < stars().length; i++) {
                if (hoverRating >= stars()[i]) {
                    setRatingLabel(starLabel[hoverRating]);
                }
            }
        }
        if (customLabel) {
            for (let i = 0; i < stars().length; i++) {
                if (hoverRating >= stars()[i]) {
                    setCustomRatingLabel(customStarLabel[hoverRating]);
                }
            }
        }
    });

    //remove hover label on mouse leave and save current index label
    const removeHoverLabel = () => {
        if (hoverLabel) {
            for (let i = 0; i < stars().length; i++) {
                if (hoverRating >= stars()[i]) {
                    setRatingLabel(starLabel[rating]);
                }
            }
        }
        if (customLabel) {
            for (let i = 0; i < stars().length; i++) {
                if (hoverRating >= stars()[i]) {
                    setCustomRatingLabel(customStarLabel[rating]);
                }
            }
        }
    };

    //make stars array function
    const stars = () => {
        let starArr = [];
        for (let i = 1; i <= count; i++) {
            starArr.push(i);
        }
        return starArr;
    };

    //render stars
    const renderStars = stars().map((index: number) => {
        return (
            <button
                key={index}
                name="star-input"
                id="star-input"
                className={`${classNames} star-btn ${disableCss}`}
                onMouseEnter={() => highlite(index)}
                onMouseLeave={() => {
                    removeHighlite();
                }}
                onClick={() => {
                    saveRating(index);
                    onChange(index); //storybook action addon
                }}
                disabled={disableHandler}
                value={index}
                role="button"
            >
                <StarIcon
                    starSize={starSize}
                    hoverRating={hoverRating}
                    rating={rating}
                    index={index}
                    customSize={customSize}
                    count={count}
                />
            </button>
        );
    });

    return (
        <div>
            {renderStars}
            <label htmlFor="star-input">
                {hoverLabel ? ratingLabel : customLabel ? customRatingLabel : ''}
            </label>
        </div>
    );
};

//default setting
Rating.defaultProps = {
    count: 5,
};

export default Rating;
