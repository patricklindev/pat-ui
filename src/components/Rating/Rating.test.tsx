import React from 'react';

import { render } from '@testing-library/react';
import Rating from './Rating';

describe('Rating', () => {
    it('should match snapshot', () => {
        const {asFragment} = render(<Rating> Snapshot Message </Rating>);
        expect(asFragment()).toMatchSnapshot();
    });
})