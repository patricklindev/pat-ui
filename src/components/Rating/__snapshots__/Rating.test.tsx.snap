// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating should match snapshot 1`] = `
<DocumentFragment>
  <div
    style="overflow: hidden; position: relative;"
  >
    <span
      class=""
      data-forhalf="★"
      data-index="0"
      style="position: relative; overflow: hidden; cursor: pointer; display: block; float: left; color: gray; font-size: 30px;"
    >
      ★
    </span>
    <span
      class=""
      data-forhalf="★"
      data-index="1"
      style="position: relative; overflow: hidden; cursor: pointer; display: block; float: left; color: gray; font-size: 30px;"
    >
      ★
    </span>
    <span
      class=""
      data-forhalf="★"
      data-index="2"
      style="position: relative; overflow: hidden; cursor: pointer; display: block; float: left; color: gray; font-size: 30px;"
    >
      ★
    </span>
    <span
      class=""
      data-forhalf="★"
      data-index="3"
      style="position: relative; overflow: hidden; cursor: pointer; display: block; float: left; color: gray; font-size: 30px;"
    >
      ★
    </span>
    <span
      class=""
      data-forhalf="★"
      data-index="4"
      style="position: relative; overflow: hidden; cursor: pointer; display: block; float: left; color: gray; font-size: 30px;"
    >
      ★
    </span>
  </div>
</DocumentFragment>
`;
